gossip/src/weighted_shuffle.rs

```diff
@@ -36,32 +37,17 @@ where
         .collect()
 }
 
-/// Returns the highest index after computing a weighted shuffle.
-/// Saves doing any sorting for O(n) max calculation.
-// TODO: Remove in favor of rand::distributions::WeightedIndex.
```

Comment:
So my comment here is actually misleading with respect to turbine usage of this function. The comment was added in https://github.com/solana-labs/solana/pull/13919 if you want to see the background context.

For all use cases which use `weighted_best` just to randomly select one index given some weights, we can replace `weighted_best` with `WeightedIndex` (and generally `WeightedIndex` has better performance unless the weights are extremely skewed).

But for turbine, we need both `broadcast` and `retransmit` shuffle nodes in the same way when they setup the fan-out tree. retransmit uses the `weighted_shuffle` function, which implements a heuristic and is not probabilistically equivalent with `WeightedIndex` (or how you would implement shuffle using `WeightedIndex`):
https://github.com/solana-labs/solana/blob/e808f34b0/gossip/src/weighted_shuffle.rs#L11-L37

Nevertheless `broadcast` has to use the same function (i.e `weighted_shuffle`), otherwise they will mismatch in how they arrange nodes on the fan-out tree which results in some nodes not getting packets.

Now, `broadcast` only cares about the root of the this tree (i.e. the very first node in `weighted_shuffle`), and `weighted_best` implements an optimization to obtain that very first node with less computation than `weighted_shuffle`. That optimization and the entire implementation of `weighted_best` is very specific and tied to `weighted_shuffle` heuristic algorithm, and so `weighted_best` cannot be replaced with `WeightedIndex` if we want to be consistent with `weighted_shuffle`.